### PR TITLE
Expose custom metrics and get rid of cAdvisor. Restructure backend files.

### PR DESCRIPTION
**Pull request by scotthallock**
### Get rid of cAdvisor and intsead expose metrics using the `prom-client` library.
  - The metrics exposed (for each Docker container) are `cpu_usage_percent` and `memory_usage_percent`
  - These metrics match the CPU % and MEM % values would would see by running `docker stats` in the terminal
  - Promethesus is configured to scrape these custom metrics on `host.docker.internal:3333/metrics`
  - The metrics are also being exposed on `localhost:3333/metrics` for development purposes
    - A separate express server was created for this 

### Restructure the backend files
  - Now there are two servers running in the backend, which are split into two files
    - extensionServer (which serves the extension frontend)
    - metricsServer (which serves the metrics to Prometheus)
  - 👉 Created a new entrypoint, `index.ts`
  - Created a `src` folder which all other `.ts` files will live in
  - Created a `constants.ts` file for important constants
  - Using the `axios` library instead of `http` to make requests (syntactic sugar)
  
 ## Next steps
   - Create Grafana dashboards for each container (by making requests to the API)
   - Listen for **events** on the Docker Engine using the API.
     - If a new container is started, get its ID, and send another request to start streaming its stats. Add these stats to the collected metrics.
   - Connect it all to the frontend with the correct Grafana iframe embed src url
  
### Metrics from 4 containers being exposed on localhost:3333/metrics
Screenshot 2023-06-28 at 5 43 04 PM

### Metrics can be queried in prometheus:
Screenshot 2023-06-28 at 5 54 11 PM

### CODE REVIEW

File: backend/index.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+metricsServer.listen(METRICS_PORT, () => {
+  console.log(`📈 Metrics are available on ${METRICS_PORT}/metrics`);
+});
+
+(async () => {
+  try {
+    // Wait 15 seconds to give the Grafana container extra time to spin up and be ready to respond
+    // to requests. This implementation can be improved later.
+    console.log('⏳ Waiting ... ');
+    await new Promise((r) => setTimeout(r, 1000 * 15));
```
Comment (scotthallock) [inline backend/index.ts:36]: @patrickv77 your idea, implemented a little bit differently here

File: backend/index.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    // Get a list of all the Docker containers
+    const response = await axios.get('/containers/json', {
+      socketPath: DOCKER_DAEMON_SOCKET_PATH,
+      params: { all: true },
+    });
+    const containers = (response.data as DockerContainer[]).map(({ Id }) => Id);
+
+    // ###########################################################
+    // # TODO: Create Grafana dashboards for each container here #
+    // ###########################################################
```
Comment (scotthallock) [inline backend/index.ts:48]: @patrickv77 this is where we need to make requests to the Grafana API to create dashboards for each container

File: backend/src/types.ts [hunk 3]
```diff
@@ -105,3 +105,18 @@ export type GrafanaPanel = {
   title: string;
   type: string;
 };
+
+export type DockerContainer = {
```
Comment (scotthallock) [inline backend/src/types.ts:109]: @patrickv77 I moved this type definition here

**Review by patrickv77 — APPROVED** [review]
Nice. Looks really clean. Liking the modularity. Nice job figuring out how to trim cAdvisor!

**Review by leejun07 — APPROVED** [review]

**Review by davidchuang5 — APPROVED** [review]

**Pull request merged**